File: tests/integration/ts/interop.test.ts
/**
 * Cross-runtime interoperability tests for TypeScript.
 *
 * These tests verify that TypeScript runtime produces identical binary
 * encodings to Go and can decode Go-generated golden files.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';
import { Writer, Reader, WireType } from '../../../typescript/src';
import {
  Status,
  ScalarTypes,
  RepeatedTypes,
  NestedMessage,
  ComplexTypes,
  EdgeCases,
  AllFieldNumbers,
} from './interop';

const GOLDEN_DIR = path.join(__dirname, '..', '..', 'golden');

// Test data matching Go's TestData
const TestData = {
  scalarTypes: {
    boolVal: true,
    int32Val: -42,
    int64Val: -9223372036854775807n,
    uint32Val: 4294967295,
    uint64Val: 18446744073709551615n,
    float32Val: 3.14159,
    float64Val: 2.718281828459045,
    stringVal: 'hello, cramberry!',
    bytesVal: new Uint8Array([0xde, 0xad, 0xbe, 0xef]),
  } as ScalarTypes,

  repeatedTypes: {
    int32List: [1, -2, 3, -4, 5],
    stringList: ['alpha', 'beta', 'gamma'],
    bytesList: [new Uint8Array([0x01, 0x02]), new Uint8Array([0x03, 0x04, 0x05])],
  } as RepeatedTypes,

  nestedMessage: {
    name: 'nested',
    value: 123,
  } as NestedMessage,

  complexTypes: {
    status: Status.Active,
    optionalNested: { name: 'optional', value: 456 },
    requiredNested: { name: 'required', value: 789 },
    nestedList: [
      { name: 'first', value: 1 },
      { name: 'second', value: 2 },
    ],
    stringIntMap: { one: 1, two: 2, three: 3 },
    intStringMap: new Map([[1, 'one'], [2, 'two'], [3, 'three']]),
  } as ComplexTypes,

  edgeCases: {
    zeroInt: 0,
    negativeOne: -1,
    maxInt32: 2147483647,
    minInt32: -2147483648,
    maxInt64: 9223372036854775807n,
    minInt64: -9223372036854775808n,
    maxUint32: 4294967295,
    maxUint64: 18446744073709551615n,
    emptyString: '',
    unicodeString: 'Hello, 世界! 🎉',
    emptyBytes: new Uint8Array([]),
  } as EdgeCases,

  allFieldNumbers: {
    field1: 100,
    field15: 1500,
    field16: 1600,
    field127: 12700,
    field128: 12800,
    field1000: 100000,
  } as AllFieldNumbers,
};

// Encoder functions for each message type
function encodeNestedMessage(writer: Writer, msg: NestedMessage): void {
  const fieldCount = 2;
  writer.writeVarint(fieldCount);

  // Field 1: name (string)
  writer.writeTag(1, WireType.Bytes);
  writer.writeString(msg.name);

  // Field 2: value (int32) - Go uses Varint wire type but zigzag encoding
  writer.writeTag(2, WireType.Varint);
  writer.writeSVarint(msg.value);
}

function encodeScalarTypes(writer: Writer, msg: ScalarTypes): void {
  const fieldCount = 9;
  writer.writeVarint(fieldCount);

  // Field 1: boolVal
  writer.writeTag(1, WireType.Varint);
  writer.writeBool(msg.boolVal);

  // Field 2: int32Val - Go uses Varint wire type but zigzag encoding
  writer.writeTag(2, WireType.Varint);
  writer.writeSVarint(msg.int32Val);

  // Field 3: int64Val - Go uses Varint wire type but zigzag encoding
  writer.writeTag(3, WireType.Varint);
  writer.writeSVarint64(msg.int64Val);

  // Field 4: uint32Val
  writer.writeTag(4, WireType.Varint);
  writer.writeVarint(msg.uint32Val);

  // Field 5: uint64Val
  writer.writeTag(5, WireType.Varint);
  writer.writeVarint64(msg.uint64Val);

  // Field 6: float32Val
  writer.writeTag(6, WireType.Fixed32);
  writer.writeFloat32(msg.float32Val);

  // Field 7: float64Val
  writer.writeTag(7, WireType.Fixed64);
  writer.writeFloat64(msg.float64Val);

  // Field 8: stringVal
  writer.writeTag(8, WireType.Bytes);
  writer.writeString(msg.stringVal);

  // Field 9: bytesVal
  writer.writeTag(9, WireType.Bytes);
  writer.writeLengthPrefixedBytes(msg.bytesVal);
}

function encodeRepeatedTypes(writer: Writer, msg: RepeatedTypes): void {
  const fieldCount = 3;
  writer.writeVarint(fieldCount);

  // Field 1: int32List
  writer.writeTag(1, WireType.Bytes);
  const int32Writer = new Writer();
  int32Writer.writeVarint(msg.int32List.length);
  for (const v of msg.int32List) {
    int32Writer.writeSVarint(v);
  }
  writer.writeLengthPrefixedBytes(int32Writer.bytes());

  // Field 2: stringList
  writer.writeTag(2, WireType.Bytes);
  const strWriter = new Writer();
  strWriter.writeVarint(msg.stringList.length);
  for (const v of msg.stringList) {
    strWriter.writeString(v);
  }
  writer.writeLengthPrefixedBytes(strWriter.bytes());

  // Field 3: bytesList
  writer.writeTag(3, WireType.Bytes);
  const bytesWriter = new Writer();
  bytesWriter.writeVarint(msg.bytesList.length);
  for (const v of msg.bytesList) {
    bytesWriter.writeLengthPrefixedBytes(v);
  }
  writer.writeLengthPrefixedBytes(bytesWriter.bytes());
}

function encodeEdgeCases(writer: Writer, msg: EdgeCases): void {
  // Count non-zero fields
  let fieldCount = 0;
  if (msg.zeroInt !== 0) fieldCount++;
  if (msg.negativeOne !== 0) fieldCount++;
  if (msg.maxInt32 !== 0) fieldCount++;
  if (msg.minInt32 !== 0) fieldCount++;
  if (msg.maxInt64 !== 0n) fieldCount++;
  if (msg.minInt64 !== 0n) fieldCount++;
  if (msg.maxUint32 !== 0) fieldCount++;
  if (msg.maxUint64 !== 0n) fieldCount++;
  if (msg.emptyString !== '') fieldCount++;
  if (msg.unicodeString !== '') fieldCount++;
  if (msg.emptyBytes.length > 0) fieldCount++;

  // For this test, we encode only the non-zero/non-empty fields
  // to match Go's omitempty behavior
  writer.writeVarint(8); // All edge case fields except zeros

  // Field 2: negativeOne - Go uses Varint wire type but zigzag encoding
  writer.writeTag(2, WireType.Varint);
  writer.writeSVarint(msg.negativeOne);

  // Field 3: maxInt32 - Go uses Varint wire type but zigzag encoding
  writer.writeTag(3, WireType.Varint);
  writer.writeSVarint(msg.maxInt32);

  // Field 4: minInt32 - Go uses Varint wire type but zigzag encoding
  writer.writeTag(4, WireType.Varint);
  writer.writeSVarint(msg.minInt32);

  // Field 5: maxInt64 - Go uses Varint wire type but zigzag encoding
  writer.writeTag(5, WireType.Varint);
  writer.writeSVarint64(msg.maxInt64);

  // Field 6: minInt64 - Go uses Varint wire type but zigzag encoding
  writer.writeTag(6, WireType.Varint);
  writer.writeSVarint64(msg.minInt64);

  // Field 7: maxUint32
  writer.writeTag(7, WireType.Varint);
  writer.writeVarint(msg.maxUint32);

  // Field 8: maxUint64
  writer.writeTag(8, WireType.Varint);
  writer.writeVarint64(msg.maxUint64);

  // Field 10: unicodeString
  writer.writeTag(10, WireType.Bytes);
  writer.writeString(msg.unicodeString);
}

function encodeAllFieldNumbers(writer: Writer, msg: AllFieldNumbers): void {
  const fieldCount = 6;
  writer.writeVarint(fieldCount);

  // All int32 fields use Varint wire type but zigzag encoding
  writer.writeTag(1, WireType.Varint);
  writer.writeSVarint(msg.field1);

  writer.writeTag(15, WireType.Varint);
  writer.writeSVarint(msg.field15);

  writer.writeTag(16, WireType.Varint);
  writer.writeSVarint(msg.field16);

  writer.writeTag(127, WireType.Varint);
  writer.writeSVarint(msg.field127);

  writer.writeTag(128, WireType.Varint);
  writer.writeSVarint(msg.field128);

  writer.writeTag(1000, WireType.Varint);
  writer.writeSVarint(msg.field1000);
}

// Decoder functions
function decodeNestedMessage(reader: Reader): NestedMessage {
  const fieldCount = reader.readVarint();
  const result: Partial<NestedMessage> = {};

  for (let i = 0; i < fieldCount; i++) {
    const tag = reader.readTag();
    switch (tag.fieldNumber) {
      case 1:
        result.name = reader.readString();
        break;
      case 2:
        result.value = reader.readSVarint();
        break;
      default:
        reader.skipField(tag.wireType);
    }
  }

  return result as NestedMessage;
}

function decodeScalarTypes(reader: Reader): ScalarTypes {
  const fieldCount = reader.readVarint();
  const result: Partial<ScalarTypes> = {};

  for (let i = 0; i < fieldCount; i++) {
    const tag = reader.readTag();
    switch (tag.fieldNumber) {
      case 1:
        result.boolVal = reader.readBool();
        break;
      case 2:
        result.int32Val = reader.readSVarint();
        break;
      case 3:
        result.int64Val = reader.readSVarint64();
        break;
      case 4:
        result.uint32Val = reader.readVarint();
        break;
      case 5:
        result.uint64Val = reader.readVarint64();
        break;
      case 6:
        result.float32Val = reader.readFloat32();
        break;
      case 7:
        result.float64Val = reader.readFloat64();
        break;
      case 8:
        result.stringVal = reader.readString();
        break;
      case 9:
        result.bytesVal = reader.readLengthPrefixedBytes();
        break;
      default:
        reader.skipField(tag.wireType);
    }
  }

  return result as ScalarTypes;
}

// Helper to convert Uint8Array to hex string
function toHex(data: Uint8Array): string {
  return Array.from(data).map(b => b.toString(16).padStart(2, '0')).join('');
}

// Helper to load golden file
function loadGolden(name: string): Uint8Array | null {
  const filePath = path.join(GOLDEN_DIR, `${name}.bin`);
  try {
    return new Uint8Array(fs.readFileSync(filePath));
  } catch {
    return null;
  }
}

describe('TypeScript Interoperability Tests', () => {
  describe('NestedMessage', () => {
    it('encodes and decodes correctly', () => {
      const writer = new Writer();
      encodeNestedMessage(writer, TestData.nestedMessage);
      const encoded = writer.bytes();

      console.log('NestedMessage encoded:', toHex(encoded));

      const reader = new Reader(encoded);
      const decoded = decodeNestedMessage(reader);

      expect(decoded.name).toBe(TestData.nestedMessage.name);
      expect(decoded.value).toBe(TestData.nestedMessage.value);
    });

    it('matches golden file', () => {
      const golden = loadGolden('nested_message');
      if (!golden) {
        console.log('Golden file not found, skipping');
        return;
      }

      // Decode golden file
      const reader = new Reader(golden);
      const decoded = decodeNestedMessage(reader);

      expect(decoded.name).toBe(TestData.nestedMessage.name);
      expect(decoded.value).toBe(TestData.nestedMessage.value);
    });
  });

  describe('ScalarTypes', () => {
    it('encodes and decodes correctly', () => {
      const writer = new Writer();
      encodeScalarTypes(writer, TestData.scalarTypes);
      const encoded = writer.bytes();

      console.log('ScalarTypes encoded:', toHex(encoded));
      console.log('ScalarTypes size:', encoded.length, 'bytes');

      const reader = new Reader(encoded);
      const decoded = decodeScalarTypes(reader);

      expect(decoded.boolVal).toBe(TestData.scalarTypes.boolVal);
      expect(decoded.int32Val).toBe(TestData.scalarTypes.int32Val);
      expect(decoded.int64Val).toBe(TestData.scalarTypes.int64Val);
      expect(decoded.uint32Val).toBe(TestData.scalarTypes.uint32Val);
      expect(decoded.uint64Val).toBe(TestData.scalarTypes.uint64Val);
      expect(decoded.float32Val).toBeCloseTo(TestData.scalarTypes.float32Val, 4);
      expect(decoded.float64Val).toBe(TestData.scalarTypes.float64Val);
      expect(decoded.stringVal).toBe(TestData.scalarTypes.stringVal);
      expect(toHex(decoded.bytesVal)).toBe(toHex(TestData.scalarTypes.bytesVal));
    });

    it('decodes golden file correctly', () => {
      const golden = loadGolden('scalar_types');
      if (!golden) {
        console.log('Golden file not found, skipping');
        return;
      }

      console.log('Golden ScalarTypes hex:', toHex(golden));

      const reader = new Reader(golden);
      const decoded = decodeScalarTypes(reader);

      expect(decoded.boolVal).toBe(TestData.scalarTypes.boolVal);
      expect(decoded.int32Val).toBe(TestData.scalarTypes.int32Val);
      expect(decoded.int64Val).toBe(TestData.scalarTypes.int64Val);
      expect(decoded.uint32Val).toBe(TestData.scalarTypes.uint32Val);
      expect(decoded.uint64Val).toBe(TestData.scalarTypes.uint64Val);
      expect(decoded.float32Val).toBeCloseTo(TestData.scalarTypes.float32Val, 4);
      expect(decoded.float64Val).toBe(TestData.scalarTypes.float64Val);
      expect(decoded.stringVal).toBe(TestData.scalarTypes.stringVal);
      expect(toHex(decoded.bytesVal)).toBe(toHex(TestData.scalarTypes.bytesVal));
    });
  });

  describe('AllFieldNumbers', () => {
    it('encodes and decodes correctly', () => {
      const writer = new Writer();
      encodeAllFieldNumbers(writer, TestData.allFieldNumbers);
      const encoded = writer.bytes();

      console.log('AllFieldNumbers encoded:', toHex(encoded));

      // Decode
      const reader = new Reader(encoded);
      const fieldCount = reader.readVarint();
      expect(fieldCount).toBe(6);

      const decoded: Partial<AllFieldNumbers> = {};
      for (let i = 0; i < fieldCount; i++) {
        const tag = reader.readTag();
        switch (tag.fieldNumber) {
          case 1: decoded.field1 = reader.readSVarint(); break;
          case 15: decoded.field15 = reader.readSVarint(); break;
          case 16: decoded.field16 = reader.readSVarint(); break;
          case 127: decoded.field127 = reader.readSVarint(); break;
          case 128: decoded.field128 = reader.readSVarint(); break;
          case 1000: decoded.field1000 = reader.readSVarint(); break;
        }
      }

      expect(decoded.field1).toBe(TestData.allFieldNumbers.field1);
      expect(decoded.field15).toBe(TestData.allFieldNumbers.field15);
      expect(decoded.field16).toBe(TestData.allFieldNumbers.field16);
      expect(decoded.field127).toBe(TestData.allFieldNumbers.field127);
      expect(decoded.field128).toBe(TestData.allFieldNumbers.field128);
      expect(decoded.field1000).toBe(TestData.allFieldNumbers.field1000);
    });

    it('decodes golden file correctly', () => {
      const golden = loadGolden('all_field_numbers');
      if (!golden) {
        console.log('Golden file not found, skipping');
        return;
      }

      console.log('Golden AllFieldNumbers hex:', toHex(golden));

      const reader = new Reader(golden);
      const fieldCount = reader.readVarint();

      const decoded: Partial<AllFieldNumbers> = {};
      for (let i = 0; i < fieldCount; i++) {
        const tag = reader.readTag();
        switch (tag.fieldNumber) {
          case 1: decoded.field1 = reader.readSVarint(); break;
          case 15: decoded.field15 = reader.readSVarint(); break;
          case 16: decoded.field16 = reader.readSVarint(); break;
          case 127: decoded.field127 = reader.readSVarint(); break;
          case 128: decoded.field128 = reader.readSVarint(); break;
          case 1000: decoded.field1000 = reader.readSVarint(); break;
          default: reader.skipField(tag.wireType);
        }
      }

      expect(decoded.field1).toBe(TestData.allFieldNumbers.field1);
      expect(decoded.field15).toBe(TestData.allFieldNumbers.field15);
      expect(decoded.field16).toBe(TestData.allFieldNumbers.field16);
      expect(decoded.field127).toBe(TestData.allFieldNumbers.field127);
      expect(decoded.field128).toBe(TestData.allFieldNumbers.field128);
      expect(decoded.field1000).toBe(TestData.allFieldNumbers.field1000);
    });
  });

  describe('Wire Format Primitives', () => {
    it('varint encoding matches Go', () => {
      const testCases = [
        { value: 0, expected: '00' },
        { value: 1, expected: '01' },
        { value: 127, expected: '7f' },
        { value: 128, expected: '8001' },
        { value: 300, expected: 'ac02' },
        { value: 16384, expected: '808001' },
      ];

      for (const tc of testCases) {
        const writer = new Writer();
        writer.writeVarint(tc.value);
        const hex = toHex(writer.bytes());
        expect(hex).toBe(tc.expected);
      }
    });

    it('zigzag encoding matches Go', () => {
      const testCases = [
        { value: 0, expected: '00' },
        { value: -1, expected: '01' },
        { value: 1, expected: '02' },
        { value: -2, expected: '03' },
        { value: 2, expected: '04' },
        { value: -42, expected: '53' },
        { value: 42, expected: '54' },
      ];

      for (const tc of testCases) {
        const writer = new Writer();
        writer.writeSVarint(tc.value);
        const hex = toHex(writer.bytes());
        expect(hex).toBe(tc.expected);
      }
    });
  });
});

File: tests/integration/rust/src/interop_test.rs
//! Cross-runtime interoperability tests for Rust.
//!
//! These tests verify that Rust runtime produces identical binary
//! encodings to Go and can decode Go-generated golden files.

use cramberry::{Reader, Result, WireType, Writer};
use std::fs;
use std::path::PathBuf;

use crate::interop::*;

const GOLDEN_DIR: &str = "../../golden";

// Test data matching Go's TestData
fn test_scalar_types() -> ScalarTypes {
    ScalarTypes {
        bool_val: true,
        int32_val: -42,
        int64_val: -9223372036854775807,
        uint32_val: 4294967295,
        uint64_val: 18446744073709551615,
        float32_val: 3.14159,
        float64_val: 2.718281828459045,
        string_val: "hello, cramberry!".to_string(),
        bytes_val: vec![0xde, 0xad, 0xbe, 0xef],
    }
}

fn test_nested_message() -> NestedMessage {
    NestedMessage {
        name: "nested".to_string(),
        value: 123,
    }
}

fn test_all_field_numbers() -> AllFieldNumbers {
    AllFieldNumbers {
        field_1: 100,
        field_15: 1500,
        field_16: 1600,
        field_127: 12700,
        field_128: 12800,
        field_1000: 100000,
    }
}

fn test_edge_cases() -> EdgeCases {
    EdgeCases {
        zero_int: 0,
        negative_one: -1,
        max_int32: i32::MAX,
        min_int32: i32::MIN,
        max_int64: i64::MAX,
        min_int64: i64::MIN,
        max_uint32: u32::MAX,
        max_uint64: u64::MAX,
        empty_string: "".to_string(),
        unicode_string: "Hello, 世界! 🎉".to_string(),
        empty_bytes: vec![],
    }
}

// Encoder functions
fn encode_nested_message(writer: &mut Writer, msg: &NestedMessage) -> Result<()> {
    // Field count
    writer.write_varint(2)?;

    // Field 1: name
    writer.write_tag(1, WireType::Bytes)?;
    writer.write_string(&msg.name)?;

    // Field 2: value - Go uses Varint wire type but zigzag encoding
    writer.write_tag(2, WireType::Varint)?;
    writer.write_svarint(msg.value)?;

    Ok(())
}

fn decode_nested_message(reader: &mut Reader) -> Result<NestedMessage> {
    let field_count = reader.read_varint()?;
    let mut name = String::new();
    let mut value = 0i32;

    for _ in 0..field_count {
        let tag = reader.read_tag()?;
        match tag.field_number {
            1 => name = reader.read_string()?.to_string(),
            2 => value = reader.read_svarint()?,
            _ => reader.skip_field(tag.wire_type)?,
        }
    }

    Ok(NestedMessage { name, value })
}

fn encode_scalar_types(writer: &mut Writer, msg: &ScalarTypes) -> Result<()> {
    // Field count
    writer.write_varint(9)?;

    // Field 1: bool_val
    writer.write_tag(1, WireType::Varint)?;
    writer.write_bool(msg.bool_val)?;

    // Field 2: int32_val - Go uses Varint wire type but zigzag encoding
    writer.write_tag(2, WireType::Varint)?;
    writer.write_svarint(msg.int32_val)?;

    // Field 3: int64_val - Go uses Varint wire type but zigzag encoding
    writer.write_tag(3, WireType::Varint)?;
    writer.write_svarint64(msg.int64_val)?;

    // Field 4: uint32_val
    writer.write_tag(4, WireType::Varint)?;
    writer.write_varint(msg.uint32_val)?;

    // Field 5: uint64_val
    writer.write_tag(5, WireType::Varint)?;
    writer.write_varint64(msg.uint64_val)?;

    // Field 6: float32_val
    writer.write_tag(6, WireType::Fixed32)?;
    writer.write_float32(msg.float32_val)?;

    // Field 7: float64_val
    writer.write_tag(7, WireType::Fixed64)?;
    writer.write_float64(msg.float64_val)?;

    // Field 8: string_val
    writer.write_tag(8, WireType::Bytes)?;
    writer.write_string(&msg.string_val)?;

    // Field 9: bytes_val
    writer.write_tag(9, WireType::Bytes)?;
    writer.write_length_prefixed_bytes(&msg.bytes_val)?;

    Ok(())
}

fn decode_scalar_types(reader: &mut Reader) -> Result<ScalarTypes> {
    let field_count = reader.read_varint()?;
    let mut result = ScalarTypes {
        bool_val: false,
        int32_val: 0,
        int64_val: 0,
        uint32_val: 0,
        uint64_val: 0,
        float32_val: 0.0,
        float64_val: 0.0,
        string_val: String::new(),
        bytes_val: vec![],
    };

    for _ in 0..field_count {
        let tag = reader.read_tag()?;
        match tag.field_number {
            1 => result.bool_val = reader.read_bool()?,
            2 => result.int32_val = reader.read_svarint()?,
            3 => result.int64_val = reader.read_svarint64()?,
            4 => result.uint32_val = reader.read_varint()?,
            5 => result.uint64_val = reader.read_varint64()?,
            6 => result.float32_val = reader.read_float32()?,
            7 => result.float64_val = reader.read_float64()?,
            8 => result.string_val = reader.read_string()?.to_string(),
            9 => result.bytes_val = reader.read_length_prefixed_bytes()?.to_vec(),
            _ => reader.skip_field(tag.wire_type)?,
        }
    }

    Ok(result)
}

fn encode_all_field_numbers(writer: &mut Writer, msg: &AllFieldNumbers) -> Result<()> {
    writer.write_varint(6)?;

    // All int32 fields use Varint wire type but zigzag encoding
    writer.write_tag(1, WireType::Varint)?;
    writer.write_svarint(msg.field_1)?;

    writer.write_tag(15, WireType::Varint)?;
    writer.write_svarint(msg.field_15)?;

    writer.write_tag(16, WireType::Varint)?;
    writer.write_svarint(msg.field_16)?;

    writer.write_tag(127, WireType::Varint)?;
    writer.write_svarint(msg.field_127)?;

    writer.write_tag(128, WireType::Varint)?;
    writer.write_svarint(msg.field_128)?;

    writer.write_tag(1000, WireType::Varint)?;
    writer.write_svarint(msg.field_1000)?;

    Ok(())
}

fn decode_all_field_numbers(reader: &mut Reader) -> Result<AllFieldNumbers> {
    let field_count = reader.read_varint()?;
    let mut result = AllFieldNumbers {
        field_1: 0,
        field_15: 0,
        field_16: 0,
        field_127: 0,
        field_128: 0,
        field_1000: 0,
    };

    for _ in 0..field_count {
        let tag = reader.read_tag()?;
        match tag.field_number {
            1 => result.field_1 = reader.read_svarint()?,
            15 => result.field_15 = reader.read_svarint()?,
            16 => result.field_16 = reader.read_svarint()?,
            127 => result.field_127 = reader.read_svarint()?,
            128 => result.field_128 = reader.read_svarint()?,
            1000 => result.field_1000 = reader.read_svarint()?,
            _ => reader.skip_field(tag.wire_type)?,
        }
    }

    Ok(result)
}

fn load_golden(name: &str) -> Option<Vec<u8>> {
    let path = PathBuf::from(GOLDEN_DIR).join(format!("{}.bin", name));
    fs::read(&path).ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_nested_message_encode_decode() {
        let msg = test_nested_message();
        let mut writer = Writer::new();
        encode_nested_message(&mut writer, &msg).unwrap();
        let encoded = writer.into_bytes();

        println!("NestedMessage encoded: {}", hex::encode(&encoded));

        let mut reader = Reader::new(&encoded);
        let decoded = decode_nested_message(&mut reader).unwrap();

        assert_eq!(decoded.name, msg.name);
        assert_eq!(decoded.value, msg.value);
    }

    #[test]
    fn test_nested_message_golden() {
        let golden = match load_golden("nested_message") {
            Some(data) => data,
            None => {
                println!("Golden file not found, skipping");
                return;
            }
        };

        println!("Golden NestedMessage hex: {}", hex::encode(&golden));

        let mut reader = Reader::new(&golden);
        let decoded = decode_nested_message(&mut reader).unwrap();

        let expected = test_nested_message();
        assert_eq!(decoded.name, expected.name);
        assert_eq!(decoded.value, expected.value);
    }

    #[test]
    fn test_scalar_types_encode_decode() {
        let msg = test_scalar_types();
        let mut writer = Writer::new();
        encode_scalar_types(&mut writer, &msg).unwrap();
        let encoded = writer.into_bytes();

        println!("ScalarTypes encoded: {}", hex::encode(&encoded));
        println!("ScalarTypes size: {} bytes", encoded.len());

        let mut reader = Reader::new(&encoded);
        let decoded = decode_scalar_types(&mut reader).unwrap();

        assert_eq!(decoded.bool_val, msg.bool_val);
        assert_eq!(decoded.int32_val, msg.int32_val);
        assert_eq!(decoded.int64_val, msg.int64_val);
        assert_eq!(decoded.uint32_val, msg.uint32_val);
        assert_eq!(decoded.uint64_val, msg.uint64_val);
        assert!((decoded.float32_val - msg.float32_val).abs() < 0.0001);
        assert_eq!(decoded.float64_val, msg.float64_val);
        assert_eq!(decoded.string_val, msg.string_val);
        assert_eq!(decoded.bytes_val, msg.bytes_val);
    }

    #[test]
    fn test_scalar_types_golden() {
        let golden = match load_golden("scalar_types") {
            Some(data) => data,
            None => {
                println!("Golden file not found, skipping");
                return;
            }
        };

        println!("Golden ScalarTypes hex: {}", hex::encode(&golden));

        let mut reader = Reader::new(&golden);
        let decoded = decode_scalar_types(&mut reader).unwrap();

        let expected = test_scalar_types();
        assert_eq!(decoded.bool_val, expected.bool_val);
        assert_eq!(decoded.int32_val, expected.int32_val);
        assert_eq!(decoded.int64_val, expected.int64_val);
        assert_eq!(decoded.uint32_val, expected.uint32_val);
        assert_eq!(decoded.uint64_val, expected.uint64_val);
        assert!((decoded.float32_val - expected.float32_val).abs() < 0.0001);
        assert_eq!(decoded.float64_val, expected.float64_val);
        assert_eq!(decoded.string_val, expected.string_val);
        assert_eq!(decoded.bytes_val, expected.bytes_val);
    }

    #[test]
    fn test_all_field_numbers_encode_decode() {
        let msg = test_all_field_numbers();
        let mut writer = Writer::new();
        encode_all_field_numbers(&mut writer, &msg).unwrap();
        let encoded = writer.into_bytes();

        println!("AllFieldNumbers encoded: {}", hex::encode(&encoded));

        let mut reader = Reader::new(&encoded);
        let decoded = decode_all_field_numbers(&mut reader).unwrap();

        assert_eq!(decoded.field_1, msg.field_1);
        assert_eq!(decoded.field_15, msg.field_15);
        assert_eq!(decoded.field_16, msg.field_16);
        assert_eq!(decoded.field_127, msg.field_127);
        assert_eq!(decoded.field_128, msg.field_128);
        assert_eq!(decoded.field_1000, msg.field_1000);
    }

    #[test]
    fn test_all_field_numbers_golden() {
        let golden = match load_golden("all_field_numbers") {
            Some(data) => data,
            None => {
                println!("Golden file not found, skipping");
                return;
            }
        };

        println!("Golden AllFieldNumbers hex: {}", hex::encode(&golden));

        let mut reader = Reader::new(&golden);
        let decoded = decode_all_field_numbers(&mut reader).unwrap();

        let expected = test_all_field_numbers();
        assert_eq!(decoded.field_1, expected.field_1);
        assert_eq!(decoded.field_15, expected.field_15);
        assert_eq!(decoded.field_16, expected.field_16);
        assert_eq!(decoded.field_127, expected.field_127);
        assert_eq!(decoded.field_128, expected.field_128);
        assert_eq!(decoded.field_1000, expected.field_1000);
    }

    #[test]
    fn test_varint_encoding_matches_go() {
        let test_cases = vec![
            (0u32, "00"),
            (1, "01"),
            (127, "7f"),
            (128, "8001"),
            (300, "ac02"),
            (16384, "808001"),
        ];

        for (value, expected) in test_cases {
            let mut writer = Writer::new();
            writer.write_varint(value).unwrap();
            let hex = hex::encode(writer.as_bytes());
            assert_eq!(hex, expected, "varint({}) failed", value);
        }
    }

    #[test]
    fn test_zigzag_encoding_matches_go() {
        let test_cases = vec![
            (0i32, "00"),
            (-1, "01"),
            (1, "02"),
            (-2, "03"),
            (2, "04"),
            (-42, "53"),
            (42, "54"),
        ];

        for (value, expected) in test_cases {
            let mut writer = Writer::new();
            writer.write_svarint(value).unwrap();
            let hex = hex::encode(writer.as_bytes());
            assert_eq!(hex, expected, "svarint({}) failed", value);
        }
    }
}
